src/metric/rank_metric.cc

```diff
@@ -451,82 +465,88 @@ struct EvalAucPR : public Metric {
-        const bst_float wt
-          = WeightPolicy::GetWeightOfInstance(info, j, group_id);
-        total_pos += wt * h_labels[j];
-        total_neg += wt * (1.0f - h_labels[j]);
-        rec.emplace_back(h_preds[j], j);
-      }
-      XGBOOST_PARALLEL_STABLE_SORT(rec.begin(), rec.end(), common::CmpFirst);
-      // we need pos > 0 && neg > 0
-      if (0.0 == total_pos || 0.0 == total_neg) {
-        auc_error += 1;
-        continue;
-      }
-      // calculate AUC
-      double tp = 0.0, prevtp = 0.0, fp = 0.0, prevfp = 0.0, h = 0.0, a = 0.0, b = 0.0;
-      for (size_t j = 0; j < rec.size(); ++j) {
-        const bst_float wt
-          = WeightPolicy::GetWeightOfSortedRecord(info, rec, j, group_id);
-        tp += wt * h_labels[rec[j].second];
-        fp += wt * (1.0f - h_labels[rec[j].second]);
-        if ((j < rec.size() - 1 && rec[j].first != rec[j + 1].first) || j  == rec.size() - 1) {
-          if (tp == prevtp) {
-            a = 1.0;
-            b = 0.0;
-          } else {
-            h = (fp - prevfp) / (tp - prevtp);
-            a = 1.0 + h;
-            b = (prevfp - h * prevtp) / total_pos;
-          }
-          if (0.0 != b) {
-            sum_auc += (tp / total_pos - prevtp / total_pos -
-                        b / a * (std::log(a * tp / total_pos + b) -
-                                 std::log(a * prevtp / total_pos + b))) / a;
-          } else {
-            sum_auc += (tp / total_pos - prevtp / total_pos) / a;
+
+    const auto &h_labels = info.labels_.ConstHostVector();
+    const auto &h_preds = preds.ConstHostVector();
+
+    #pragma omp parallel reduction(+:sum_auc, auc_error) if (ngroups > 1)
+    {
+      // Each thread works on a distinct group and sorts the predictions in that group
+      PredIndPairContainer rec;
+      #pragma omp for schedule(static)
+      for (bst_omp_uint group_id = 0; group_id < ngroups; ++group_id) {
+        double total_pos = 0.0;
+        double total_neg = 0.0;
+        // Same thread can work on multiple groups one after another; hence, resize
+        // the predictions array based on the current group
+        rec.resize(gptr[group_id + 1] - gptr[group_id]);
+        #pragma omp parallel for schedule(static) reduction(+:total_pos, total_neg) \
+          if (!omp_in_parallel())  // NOLINT
+        for (bst_omp_uint j = gptr[group_id]; j < gptr[group_id + 1]; ++j) {
+          const bst_float wt = WeightPolicy::GetWeightOfInstance(info, j, group_id);
+          total_pos += wt * h_labels[j];
+          total_neg += wt * (1.0f - h_labels[j]);
+          rec[j - gptr[group_id]] = {h_preds[j], j};
+        }
+
+        if (omp_in_parallel()) {
+          std::stable_sort(rec.begin(), rec.end(), common::CmpFirst);
+        } else {
+          XGBOOST_PARALLEL_SORT(rec.begin(), rec.end(), common::CmpFirst);
+        }
+
+        // we need pos > 0 && neg > 0
+        if (total_pos <= 0.0 || total_neg <= 0.0) {
```

Comment:
this is the fix.

Reply:
So the fix is trigger a fatal error when only one label type exists locally? e.g. in a distributed partition.

Reply:
Nvm looks like it just ignores the partition.

Reply:
if labels aren't diverse across all groups even in a non distributed environment.